rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRACDIExtension.java

```diff
@@ -0,0 +1,249 @@
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.spi.AfterBeanDiscovery;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.util.AnnotationLiteral;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.HEAD;
+import javax.ws.rs.OPTIONS;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * This CDI extension collects all LRA participants that contain
+ * one or more non-JAX-RS participant methods. The collected classes are stored
+ * in {@link LRAParticipantRegistry}.
+ */
+public class LRACDIExtension implements Extension {
+
+    private ClassPathIndexer classPathIndexer = new ClassPathIndexer();
+    private final Map<String, LRAParticipant> participants = new HashMap<>();
+
+    public void observe(@Observes AfterBeanDiscovery event, BeanManager beanManager) throws IOException, ClassNotFoundException {
+        Index index = classPathIndexer.createIndex();
+
+        List<AnnotationInstance> annotations = index.getAnnotations(DotName.createSimple("javax.ws.rs.Path"));
+
+        for (AnnotationInstance annotation : annotations) {
+            ClassInfo classInfo;
+            AnnotationTarget target = annotation.target();
+
+            if (target.kind().equals(AnnotationTarget.Kind.CLASS)) {
+                classInfo = target.asClass();
+            } else if (target.kind().equals(AnnotationTarget.Kind.METHOD)) {
+                classInfo = target.asMethod().declaringClass();
+            } else {
+                continue;
+            }
+
+            Class<?> clazz = getClass().getClassLoader().loadClass(classInfo.name().toString());
+
+            LRAParticipant participant = processJavaClass(clazz);
+            if (participant != null) {
+                participants.put(participant.getId().getName(), participant);
+                Set<Bean<?>> participantBeans = beanManager.getBeans(clazz, new AnnotationLiteral<Any>() {});
+                if (participantBeans.isEmpty()) {
+                    // resource is not registered as managed bean so register a custom managed instance
+                    try {
+                        participant.setInstance(clazz.newInstance());
+                    } catch (InstantiationException | IllegalAccessException e) {
+                        LRALogger.i18NLogger.error_cannotProcessParticipant(e);
+                    }
+                }
+            }
+        }
+
+        event.addBean()
```

Comment:
We discussed this in the Narayana chat room and you said you were going to do more investigation.

The new non JAX-RS functionality does not require CDI and it is easy to implement the functionality without requiring CDI. Your PR is now restricting where we can use our implementation and there is no benefit to mandating CDI.

Reply:
We will need to bring the dependency on CDI because this class is a CDI extension. Is there any other way how I can catch the initialization of the underlying application in Narayana, please?

Reply:
There are at least a couple of JAX-RS specific approaches that I am aware of:

1) implement a javax.servlet.ServletContextListener
2) implement a JAX-RS dynamic feature which will be processed at deployment time.

BTW you never said that you were using CDI as a workaround for hooking into the application deployment lifecycle. So if that is the reason then okay we can go with CDI (and can rework it if there is ever a need to use our LRA implementation in a JAX-RS only environment).

Reply:
@mmusgrov thank you, I would like to keep CDI approach because it's easy to respect user decisions if users define CDI annotations themselves.